tools/ci/java-ci-tools/src/main/java/org/apache/flink/tools/ci/licensecheck/LicenseChecker.java

```diff
@@ -0,0 +1,320 @@
+ */
+
+package org.apache.flink.tools.ci.licensecheck;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.StringReader;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Scanner;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * Utility class checking for proper NOTICE files based on the maven build output.
+ */
+public class LicenseChecker {
+	// ---------------------------------------- Launcher ---------------------------------------- //
+
+	private static final Logger LOG = LoggerFactory.getLogger(LicenseChecker.class);
+
+	public static void main(String[] args) throws IOException {
+		if (args.length < 2) {
+			System.out.println("Usage: LicenseChecker <pathMavenBuildOutput> <pathFlinkRoot>");
+			System.exit(1);
+		}
+		LOG.warn("THIS UTILITY IS ONLY CHECKING FOR COMMON LICENSING MISTAKES. A MANUAL CHECK OF THE NOTICE FILES, DEPLOYED ARTIFACTS, ETC. IS STILL NEEDED!");
+
+		LicenseChecker checker = new LicenseChecker();
+		int severeIssueCount = checker.run(new File(args[0]), new File(args[1]));
+
+		if (severeIssueCount > 0) {
+			LOG.warn("Found a total of {} severe license issues", severeIssueCount);
+
+			System.exit(1);
+		}
+		LOG.info("License check completed without severe issues.");
+	}
+
+	// ---------------------------------------- License Checker ---------------------------------------- //
+
+	private static final List<String> MODULES_SKIPPING_DEPLOYMENT = loadFromResources("modules-skipping-deployment.modulelist");
+
+	private static final List<String> MODULES_DEFINING_EXCESS_DEPENDENCIES = loadFromResources("modules-defining-excess-dependencies.modulelist");
+
+	// pattern for maven shade plugin
+	private static final Pattern SHADE_NEXT_MODULE_PATTERN = Pattern.compile(".*:shade \\((shade-flink|default)\\) @ ([^ _]+)(_[0-9.]+)? --.*");
+	private static final Pattern SHADE_INCLUDE_MODULE_PATTERN = Pattern.compile(".*Including ([^:]+):([^:]+):jar:([^ ]+) in the shaded jar");
+
+	// pattern for maven-dependency-plugin copyied dependencies
+	private static final Pattern DEPENDENCY_COPY_NEXT_MODULE_PATTERN = Pattern.compile(".*maven-dependency-plugin:3.1.1:copy \\([^)]+\\) @ ([^ _]+)(_[0-9.]+)? --.*");
+	private static final Pattern DEPENDENCY_COPY_INCLUDE_MODULE_PATTERN = Pattern.compile(".*Configured Artifact: ([^:]+):([^:]+):([^:]+):jar.*");
+
+	private static final Pattern NOTICE_DEPENDENCY_PATTERN = Pattern.compile("- ([^ :]+):([^:]+):([^ ]+)($| )|.*bundles \"([^:]+):([^:]+):([^\"]+)\".*");
```

Comment:
I'd include an example for both pattern; it tends to make things a lot easier to understand.